### PR TITLE
[Experiment] Load Braze Web SDK

### PR DESCRIPTION
**Pull request by tjmw**
## What does this change?

Load the Braze Web SDK lazily for logged in users and show a default message.

## Does this change need to be reproduced in dotcom-rendering ?

- [ ] No
- [x] Yes (please indicate your plans for DCR Implementation: PR coming soon

## Screenshots

Screenshot 2020-07-14 at 16 56 06

## What is the value of this and can you measure success?

Verifying that the mechanism we're using the load the Braze Web SDK is acceptable to teams involved.

### Does this affect other platforms?

- [ ] AMP 
- [ ] Apps
- [ ] Other (please specify)

### Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

- [x] No
- [ ] Yes (please give details)

### Does this change break ad-free?

- [x] No
- [ ] It did, but tests caught it and I fixed it
- [ ] It did, but there was no test coverage so I added that then fixed it

### Does this change update the version of CAPI we're using?

- [x] No, all the existing database files are just fine
- [ ] Yes, and I have [re-run all the tests locally and checked in all the updated data/database/xyz files](https://github.com/guardian/frontend/blob/master/docs/03-dev-howtos/15-updating-test-database.md)

### Accessibility test checklist

- [ ] [Tested with screen reader](https://accessibility.gutools.co.uk/testing/web/screen-readers/)
- [ ] [Navigable with keyboard](https://accessibility.gutools.co.uk/testing/web/keyboard-navigation/)
- [ ] [Colour contrast passed](https://accessibility.gutools.co.uk/testing/web/colour-contrast/)

### Tested

- [x] Locally
- [ ] On CODE (optional)

### CODE REVIEW

File: static/src/javascripts/projects/commercial/modules/braze.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            appboy.initialize(apiKey, {
+                enableLogging: true,
+                noCookies: true,
+                baseUrl: 'https://sdk.fra-01.braze.eu/api/v3',
+                enableHtmlInAppMessages: true
+            });
+
+            const f = function (configuration) {
+                console.log(configuration);
+                appboy.display.showInAppMessage(configuration);
```
Comment (tjmw) [inline static/src/javascripts/projects/commercial/modules/braze.js:48]: Note that we're planning on doing something custom to actually show the message to the user.

File: static/src/javascripts/projects/commercial/modules/braze.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    })
+
+export const init = (): Promise<any> => {
+    const brazeSwitch = config.get('switches.brazeSwitch');
+    if (!brazeSwitch) return Promise.resolve();
+
+    const apiKey = config.get('page.brazeApiKey');
+    if (!apiKey) return Promise.reject(new Error('Braze API key not set.'));
+
+    const dependencies = [getBrazeUuid(),hasRequiredConsents()]
```
Comment (tjmw) [inline static/src/javascripts/projects/commercial/modules/braze.js:33]: The Braze UUID comes from a request to `idapi/user/me` from the browser. Requests to that endpoint are memoized, so the call here doesn't actually result in any additional API requests being made, since it seems we're already calling that endpoint on every page. Perhaps we could optimize that, but it feels out of scope for this PR.

**Comment by tjmw** [comment]
Hey @gtrufitt this is an early first pass at loading the Braze Web SDK on frontend. There's no plan to merge this as is but I'm interested if you've got any early feedback on the approach, which is basically to lazily load the SDK once we know that the user is signed-in (and so has a Braze UUID) and has given the appropriate consents via the CMP. The plan is to show custom messaging (so render the message ourselves, rather than let Braze do it). We've just made the decision that the message format will be a banner, so that does imply changing things to integrate with the banner picker, but we'll keep the lazy loading mechanism.

File: static/src/javascripts/projects/commercial/modules/braze.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+                resolve(user.privateFields.brazeUuid)
+            }
+        })
+    })
+
+const hasRequiredConsents = (): Promise<void> =>
+    new Promise(resolve => {
+        onIabConsentNotification(state => {
+            if (state[1] && state[2] && state[3] && state[4] && state[5]) {
+                resolve()
```
Comment (gtrufitt) [inline static/src/javascripts/projects/commercial/modules/braze.js:21]: I guess you want to reject this promise otherwise it might just never resolve?

**Review by gtrufitt — COMMENTED** [review]
This looks pretty good to me, I think the later you can lazy load that third party the better, so really nice pattern to handle the dependencies like this.

File: static/src/javascripts/projects/commercial/modules/braze.js (hunk 3 above)
Comment (mxdvl) [reply static/src/javascripts/projects/commercial/modules/braze.js:21]: A quick note to mention that this does not currently support CCPA or the (upcoming) TCFv2 framework. Please let us know if you require help to set this up. Some PR that might be of interest below:
- #22819 
- #22818

Comment (tjmw) [reply static/src/javascripts/projects/commercial/modules/braze.js:21]: @gtrufitt good spot! It was actually intentional that we're not rejecting here (my thinking was that it's not problematic/exceptional if the consents aren't given, but we can't proceed any further if that's the case, hence not resolving). I'd hadn't occurred to me that having an un-(resolved|rejected) promise hanging around might be problematic, I'll have a read of that thread! Having thought about it though, I can see this could be a bit weird/confusing for someone reading this code in future since it's definitely a bit unconventional, so rejecting if the consents aren't given probably a better approach all round.

Comment (tjmw) [reply static/src/javascripts/projects/commercial/modules/braze.js:21]: @mxdvl Ah yep it's definitely on my radar and this PR won't be merged until we support CCPA/TCFv2. Thanks for the PR links, they look useful. Having taken a quick look it seems like listening on consent changes is the same but we can detect CCPA from `state.ccpa` in the callback?

Comment (mxdvl) [reply static/src/javascripts/projects/commercial/modules/braze.js:21]: @tjmw that's mostly right, but as of next week there will only be the new CCPA/TCFv2 interface.

For CCPA, the in the callback on `onConsentChange`, you need to check `state.ccpa` exists/is set. Then you can confirm `state.ccpa.doNotSell === false` and resolve.

Comment (mxdvl) [reply static/src/javascripts/projects/commercial/modules/braze.js:21]: We are planning to officially release the CMP v4 next Wednesday—12 August—and will be on call to assist if need be!

**Comment by tjmw** [comment]
Closing this in favour of #22883 (which includes the same work + some additional stuff).
